File: paper/replication/validate_geoxgboost.py
#!/usr/bin/env python3
"""
Validate smelt-ml's GeoXGBoost against Grekousis' original geoxgboost package.

Runs both implementations on King County housing data (1K samples) with
matched hyperparameters, comparing RMSE and per-point predictions.

Output: validation table suitable for the JSS paper.
"""

import numpy as np
import pandas as pd
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
import time
import json
import sys
import os

# ── Load data ──────────────────────────────────────────────────────────

DATA_PATH = os.path.join(os.path.dirname(__file__), "..", "..", "data", "king_county_1k.csv")
df = pd.read_csv(DATA_PATH)

target_col = "log_price"
coord_cols = ["lat", "long"]
feature_cols = [c for c in df.columns if c not in [target_col] + coord_cols]

X = df[feature_cols]
y = df[target_col]
coords = df[coord_cols]

# Fixed train/test split (80/20, seed=42)
indices = np.arange(len(df))
train_idx, test_idx = train_test_split(indices, test_size=0.2, random_state=42)

X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]
coords_train, coords_test = coords.iloc[train_idx], coords.iloc[test_idx]

print(f"Data: {len(df)} samples, {len(feature_cols)} features")
print(f"Train: {len(train_idx)}, Test: {len(test_idx)}")
print(f"Features: {feature_cols}")
print()

# ── Matched hyperparameters ────────────────────────────────────────────

BANDWIDTH = 30
N_ESTIMATORS = 100
MAX_DEPTH = 6
LEARNING_RATE = 0.3
REG_LAMBDA = 1.0
SEED = 42

xgb_params = {
    "n_estimators": N_ESTIMATORS,
    "max_depth": MAX_DEPTH,
    "learning_rate": LEARNING_RATE,
    "reg_lambda": REG_LAMBDA,
    "random_state": SEED,
    "n_jobs": 1,
    "verbosity": 0,
}

# ── 1. Run Grekousis' geoxgboost ───────────────────────────────────────

print("=" * 60)
print("Running Grekousis geoxgboost (Python original)")
print("=" * 60)

from geoxgboost import gxgb, predict_gxgb

# geoxgboost expects DataFrames; params as dict for XGBoost
gxgb_params = {
    "n_estimators": N_ESTIMATORS,
    "max_depth": MAX_DEPTH,
    "learning_rate": LEARNING_RATE,
    "reg_lambda": REG_LAMBDA,
    "reg_alpha": 0,
    "random_state": SEED,
    "n_jobs": 1,
    "verbosity": 0,
}

t0 = time.time()
result = gxgb(
    X=X_train.reset_index(drop=True),
    y=y_train.reset_index(drop=True),
    Coords=coords_train.reset_index(drop=True),
    params=gxgb_params,
    bw=BANDWIDTH,
    Kernel="Adaptive",
    alpha_wt_type="varying",
    alpha_wt=1,
    test_size=0.2,
    seed=SEED,
    n_splits=5,
    path_save=False,
)
gxgb_train_time = time.time() - t0

# Extract training predictions
train_preds_df = result["Prediction"]
print(f"\nTrain time: {gxgb_train_time:.2f}s")
print(f"Result keys: {list(result.keys())}")
print(f"Stats:\n{result['Stats']}")

# Predict on test set
t0 = time.time()
test_result = predict_gxgb(
    DataPredict=X_test.reset_index(drop=True),
    CoordsPredict=coords_test.reset_index(drop=True),
    Coords=coords_train.reset_index(drop=True),
    Output_GXGB_LocalModel=result,
    alpha_wt_type="varying",
    path_save=False,
)
gxgb_predict_time = time.time() - t0

# Extract test predictions
if isinstance(test_result, pd.DataFrame):
    gxgb_test_preds = test_result.iloc[:, -1].values  # last column is usually prediction
    print(f"\nTest prediction columns: {list(test_result.columns)}")
elif isinstance(test_result, dict):
    print(f"\nTest prediction keys: {list(test_result.keys())}")
    if "Prediction" in test_result:
        pred_df = test_result["Prediction"]
        gxgb_test_preds = pred_df.iloc[:, -1].values
    else:
        gxgb_test_preds = np.array(list(test_result.values())[0])
else:
    gxgb_test_preds = np.array(test_result)

gxgb_test_rmse = np.sqrt(mean_squared_error(y_test, gxgb_test_preds))
gxgb_test_r2 = r2_score(y_test, gxgb_test_preds)

print(f"\ngeoxgboost Test RMSE: {gxgb_test_rmse:.4f}")
print(f"geoxgboost Test R²:   {gxgb_test_r2:.4f}")
print(f"Predict time: {gxgb_predict_time:.2f}s")

# ── 2. Run standard XGBoost (baseline) ─────────────────────────────────

print()
print("=" * 60)
print("Running standard XGBoost (baseline)")
print("=" * 60)

import xgboost as xgb

model_xgb = xgb.XGBRegressor(**xgb_params)
t0 = time.time()
model_xgb.fit(X_train, y_train)
xgb_train_time = time.time() - t0

xgb_test_preds = model_xgb.predict(X_test)
xgb_test_rmse = np.sqrt(mean_squared_error(y_test, xgb_test_preds))
xgb_test_r2 = r2_score(y_test, xgb_test_preds)

print(f"XGBoost Test RMSE: {xgb_test_rmse:.4f}")
print(f"XGBoost Test R²:   {xgb_test_r2:.4f}")
print(f"Train time: {xgb_train_time:.2f}s")

# ── 3. Summary ─────────────────────────────────────────────────────────

print()
print("=" * 60)
print("VALIDATION SUMMARY")
print("=" * 60)
print(f"{'Method':<25} {'RMSE':>8} {'R²':>8} {'Train(s)':>10}")
print("-" * 55)
print(f"{'XGBoost (baseline)':<25} {xgb_test_rmse:>8.4f} {xgb_test_r2:>8.4f} {xgb_train_time:>10.2f}")
print(f"{'geoxgboost (Grekousis)':<25} {gxgb_test_rmse:>8.4f} {gxgb_test_r2:>8.4f} {gxgb_train_time:>10.2f}")
print()
print("Save train/test indices and geoxgboost predictions for Rust comparison:")

# Save for Rust comparison
np.savetxt(
    os.path.join(os.path.dirname(__file__), "geoxgb_train_idx.csv"),
    train_idx, fmt="%d"
)
np.savetxt(
    os.path.join(os.path.dirname(__file__), "geoxgb_test_idx.csv"),
    test_idx, fmt="%d"
)
np.savetxt(
    os.path.join(os.path.dirname(__file__), "geoxgb_test_preds.csv"),
    gxgb_test_preds, fmt="%.6f"
)
np.savetxt(
    os.path.join(os.path.dirname(__file__), "xgb_test_preds.csv"),
    xgb_test_preds, fmt="%.6f"
)

# Save full results as JSON
results = {
    "dataset": "king_county_1k",
    "n_train": len(train_idx),
    "n_test": len(test_idx),
    "n_features": len(feature_cols),
    "bandwidth": BANDWIDTH,
    "n_estimators": N_ESTIMATORS,
    "max_depth": MAX_DEPTH,
    "learning_rate": LEARNING_RATE,
    "seed": SEED,
    "xgboost": {
        "rmse": round(float(xgb_test_rmse), 4),
        "r2": round(float(xgb_test_r2), 4),
        "train_time_s": round(xgb_train_time, 2),
    },
    "geoxgboost_python": {
        "rmse": round(float(gxgb_test_rmse), 4),
        "r2": round(float(gxgb_test_r2), 4),
        "train_time_s": round(gxgb_train_time, 2),
    },
}

with open(os.path.join(os.path.dirname(__file__), "geoxgb_validation.json"), "w") as f:
    json.dump(results, f, indent=2)

print("\nFiles saved in paper/replication/:")
print("  geoxgb_train_idx.csv, geoxgb_test_idx.csv")
print("  geoxgb_test_preds.csv, xgb_test_preds.csv")
print("  geoxgb_validation.json")
print("\nNext: run the Rust GeoXGBoost example with the same indices to compare.")

File: examples/validate_geoxgboost.rs
//! Validate smelt-ml GeoXGBoost against the Python geoxgboost package.
//!
//! Reads the same data and train/test indices produced by the Python script,
//! runs GeoXGBoost with matched hyperparameters, and outputs RMSE for comparison.
//!
//! Run with: cargo run --release --example validate_geoxgboost

use ndarray::Axis;
use smelt_ml::data::CsvLoader;
use smelt_ml::learner::TrainedModel;
use smelt_ml::prelude::*;
use std::fs;
use std::time::Instant;

fn read_indices(path: &str) -> Vec<usize> {
    fs::read_to_string(path)
        .unwrap_or_else(|_| panic!("Cannot read {path}. Run validate_geoxgboost.py first."))
        .lines()
        .filter(|l| !l.is_empty())
        .map(|l| l.trim().parse::<usize>().unwrap())
        .collect()
}

fn rmse(predicted: &[f64], actual: &[f64]) -> f64 {
    let n = predicted.len() as f64;
    let ss: f64 = predicted
        .iter()
        .zip(actual)
        .map(|(p, a)| (p - a).powi(2))
        .sum();
    (ss / n).sqrt()
}

fn r2(predicted: &[f64], actual: &[f64]) -> f64 {
    let mean = actual.iter().sum::<f64>() / actual.len() as f64;
    let ss_res: f64 = predicted
        .iter()
        .zip(actual)
        .map(|(p, a)| (p - a).powi(2))
        .sum();
    let ss_tot: f64 = actual.iter().map(|a| (a - mean).powi(2)).sum();
    1.0 - ss_res / ss_tot
}

fn main() {
    println!("================================================================");
    println!("  Validate smelt-ml GeoXGBoost vs geoxgboost (Python)");
    println!("================================================================\n");

    // ── Load data ──────────────────────────────────────────────────────
    let task = CsvLoader::from_path("data/king_county_1k.csv")
        .target("log_price")
        .load_regress()
        .expect("Failed to load King County dataset");

    let feature_names = task.feature_names();
    let lat_idx = feature_names.iter().position(|n| n == "lat").unwrap();
    let lon_idx = feature_names.iter().position(|n| n == "long").unwrap();

    let coords: Vec<(f64, f64)> = (0..task.n_samples())
        .map(|i| (task.features()[[i, lat_idx]], task.features()[[i, lon_idx]]))
        .collect();

    // Features WITHOUT coordinates (matches Python script)
    let non_coord_cols: Vec<usize> = (0..task.features().ncols())
        .filter(|&j| j != lat_idx && j != lon_idx)
        .collect();
    let features = task.features().select(Axis(1), &non_coord_cols).to_owned();
    let target = task.target().to_vec();
    let n_features = features.ncols();

    println!("Loaded: {} samples, {} features (excl. coords)", target.len(), n_features);

    // ── Load train/test indices from Python ────────────────────────────
    let train_idx = read_indices("paper/replication/geoxgb_train_idx.csv");
    let test_idx = read_indices("paper/replication/geoxgb_test_idx.csv");

    println!("Train: {}, Test: {} (indices from Python)\n", train_idx.len(), test_idx.len());

    let tr_feat = features.select(Axis(0), &train_idx).to_owned();
    let tr_tgt: Vec<f64> = train_idx.iter().map(|&i| target[i]).collect();
    let te_feat = features.select(Axis(0), &test_idx).to_owned();
    let te_tgt: Vec<f64> = test_idx.iter().map(|&i| target[i]).collect();
    let tr_coords: Vec<(f64, f64)> = train_idx.iter().map(|&i| coords[i]).collect();

    let tr_task = RegressionTask::new("train", tr_feat.clone(), tr_tgt.clone()).unwrap();

    // ── Matched hyperparameters ────────────────────────────────────────
    // Must match validate_geoxgboost.py exactly
    let bandwidth = 30;
    let n_estimators = 100;
    let max_depth = 6;
    let learning_rate = 0.3;
    let lambda = 1.0;
    let seed = 42;

    // ── 1. Standard XGBoost (baseline) ─────────────────────────────────
    println!("─── Standard XGBoost (baseline) ───");
    let mut xgb = XGBoost::new()
        .with_n_estimators(n_estimators)
        .with_max_depth(max_depth)
        .with_learning_rate(learning_rate)
        .with_lambda(lambda)
        .with_seed(seed);

    let t0 = Instant::now();
    let xgb_model = xgb.train_regress(&tr_task).unwrap();
    let xgb_train_time = t0.elapsed();

    let xgb_pred = xgb_model.predict(&te_feat).unwrap();
    let xgb_vals = match &xgb_pred {
        Prediction::Regression { predicted, .. } => predicted.clone(),
        _ => panic!("Expected regression"),
    };

    let xgb_rmse = rmse(&xgb_vals, &te_tgt);
    let xgb_r2 = r2(&xgb_vals, &te_tgt);
    println!("RMSE: {:.4}", xgb_rmse);
    println!("R²:   {:.4}", xgb_r2);
    println!("Time: {:.2}s\n", xgb_train_time.as_secs_f64());

    // ── 2. GeoXGBoost ──────────────────────────────────────────────────
    println!("─── GeoXGBoost (smelt-ml) ───");
    // alpha=1.0 → pure local prediction (matches geoxgboost Python defaults:
    // alpha_wt=1, alpha_wt_type='varying' → 100% local, no blending)
    let mut gxgb = GeoXGBoost::new(tr_coords.clone())
        .with_bandwidth(bandwidth)
        .with_n_estimators(n_estimators)
        .with_max_depth(max_depth)
        .with_learning_rate(learning_rate)
        .with_lambda(lambda)
        .with_alpha(1.0)
        .with_seed(seed);

    let t0 = Instant::now();
    // Use train_geo() for concrete type with predict_spatial()
    let gxgb_model = gxgb.train_geo(&tr_task).unwrap();
    let gxgb_train_time = t0.elapsed();

    // In-sample predictions (local models, via predict_spatial with the
    // training coords — predict() alone is global-only by design)
    let gxgb_train_pred = gxgb_model.predict_spatial(&tr_feat, &tr_coords).unwrap();
    let gxgb_train_vals = match &gxgb_train_pred {
        Prediction::Regression { predicted, .. } => predicted.clone(),
        _ => panic!("Expected regression"),
    };
    let gxgb_train_rmse = rmse(&gxgb_train_vals, &tr_tgt);
    let gxgb_train_r2 = r2(&gxgb_train_vals, &tr_tgt);

    // Out-of-sample predictions with spatial nearest-neighbor lookup
    let te_coords: Vec<(f64, f64)> = test_idx.iter().map(|&i| coords[i]).collect();
    let gxgb_pred = gxgb_model.predict_spatial(&te_feat, &te_coords).unwrap();
    let gxgb_vals = match &gxgb_pred {
        Prediction::Regression { predicted, .. } => predicted.clone(),
        _ => panic!("Expected regression"),
    };

    let gxgb_rmse = rmse(&gxgb_vals, &te_tgt);
    let gxgb_r2 = r2(&gxgb_vals, &te_tgt);

    println!("In-sample  RMSE: {:.4}, R²: {:.4}", gxgb_train_rmse, gxgb_train_r2);
    println!("Out-sample RMSE: {:.4}, R²: {:.4}", gxgb_rmse, gxgb_r2);
    println!("Time: {:.2}s\n", gxgb_train_time.as_secs_f64());

    // ── Save predictions for comparison ────────────────────────────────
    let pred_str: String = gxgb_vals
        .iter()
        .map(|v| format!("{:.6}", v))
        .collect::<Vec<_>>()
        .join("\n");
    fs::write("paper/replication/smelt_geoxgb_test_preds.csv", &pred_str).unwrap();

    let xgb_pred_str: String = xgb_vals
        .iter()
        .map(|v| format!("{:.6}", v))
        .collect::<Vec<_>>()
        .join("\n");
    fs::write("paper/replication/smelt_xgb_test_preds.csv", &xgb_pred_str).unwrap();

    // ── Load Python predictions for direct comparison ──────────────────
    let py_geoxgb: Vec<f64> = fs::read_to_string("paper/replication/geoxgb_test_preds.csv")
        .unwrap_or_default()
        .lines()
        .filter(|l| !l.is_empty())
        .map(|l| l.trim().parse::<f64>().unwrap_or(0.0))
        .collect();

    let py_xgb: Vec<f64> = fs::read_to_string("paper/replication/xgb_test_preds.csv")
        .unwrap_or_default()
        .lines()
        .filter(|l| !l.is_empty())
        .map(|l| l.trim().parse::<f64>().unwrap_or(0.0))
        .collect();

    println!("═══════════════════════════════════════════════════════════");
    println!("  VALIDATION SUMMARY");
    println!("═══════════════════════════════════════════════════════════");
    println!("{:<30} {:>8} {:>8}", "Method", "RMSE", "R²");
    println!("{}", "-".repeat(50));
    println!("{:<30} {:>8.4} {:>8.4}", "XGBoost (smelt-ml)", xgb_rmse, xgb_r2);
    println!("{:<30} {:>8.4} {:>8.4}", "GeoXGBoost (smelt-ml)", gxgb_rmse, gxgb_r2);

    if !py_xgb.is_empty() {
        let py_xgb_rmse = rmse(&py_xgb, &te_tgt);
        let py_xgb_r2 = r2(&py_xgb, &te_tgt);
        println!("{:<30} {:>8.4} {:>8.4}", "XGBoost (official C++)", py_xgb_rmse, py_xgb_r2);
    }
    if !py_geoxgb.is_empty() {
        let py_geoxgb_rmse = rmse(&py_geoxgb, &te_tgt);
        let py_geoxgb_r2 = r2(&py_geoxgb, &te_tgt);
        println!("{:<30} {:>8.4} {:>8.4}", "geoxgboost (Grekousis)", py_geoxgb_rmse, py_geoxgb_r2);
    }

    // Per-point correlation between implementations
    if !py_geoxgb.is_empty() && py_geoxgb.len() == gxgb_vals.len() {
        let pred_diff: Vec<f64> = gxgb_vals
            .iter()
            .zip(&py_geoxgb)
            .map(|(a, b)| (a - b).abs())
            .collect();
        let mean_diff: f64 = pred_diff.iter().sum::<f64>() / pred_diff.len() as f64;
        let max_diff: f64 = pred_diff.iter().cloned().fold(0.0_f64, f64::max);
        println!("\n─── Per-point prediction difference (GeoXGBoost) ───");
        println!("Mean |smelt - geoxgboost|: {:.4}", mean_diff);
        println!("Max  |smelt - geoxgboost|: {:.4}", max_diff);
    }

    println!("\nDone. Predictions saved to paper/replication/smelt_*.csv");
}
